### ISSUE TITLE
Postgres: ability to set schema

### ISSUE THREAD

**Issue by rogierlommers** [spotweb/spotweb#463]
Please see [this](https://stackoverflow.com/questions/34590761/how-to-set-the-connection-with-a-schema-in-a-postgresql-db-with-pdo-php) github issue. It would be great to be able to set the schema. So instead of using the "public" schema, specify another schema during installation.

This is not already part of the features of spotweb, right?

**Comment by mesa57** [comment]
Could you eleborate on the reason why you would want to have that ?

**Comment by rogierlommers** [comment]
Well, I'm running spotweb on my home server, together with a few other services (RSS aggregation, etherpad, etc). I don't want to run three postgres instances, I prefer to have one. Therefore being able to assign a schema to each services would be great.

**Comment by mesa57** [comment]
You can, because spotweb uses its own database.
Schema's are per database and only give a further partition of the database.

**Comment by rogierlommers** [comment]
I don't understand what you're saying; I currently have one database. This database has three schema's. I want spotweb to make use of one specific schema. I don't think that's possible now.

**Comment by rogierlommers** [comment]
<img width="744" alt="Screenshot 2020-02-01 at 19 41 10" src="https://user-images.githubusercontent.com/3907131/73597311-d86a2780-452a-11ea-8d52-83bafaf1438f.png">

adding the following line to `lib/dbeng/dbeng_pdo_pgsql.php` does the job. It's not very hard, but to have this "clean" as an optional property... here is my PHP knowledge lacking :). Can someone help?

**Comment by mesa57** [comment]
Of course adding that line would be easy.
But including it in the configuration is an other story.

**Comment by Sweepr** [comment]
You can also set the default search_path per database user and in that case the above statement becomes redundant.

ALTER USER user SET search_path TO accountschema;

Is this what you want?

**Comment by rogierlommers** [comment]
Hey @Sweepr , thanks for the suggestion. But it's for my home server. I have one database with one user running in a container. I want to use this database for three things. One is Spotweb :).

So setting the schema should be inside the app layer, not on user level.

**Comment by rogierlommers** [comment]
Little kick :). I would still love someone who added this as a property in the config file.

**Comment by Sweepr** [comment]
I have created a new branch called: https://github.com/spotweb/spotweb/tree/postgres please check this branch out if this works for you.

Please be aware that this is work in progress and may contain error's and mistake's.

I tried my best to get things working but at this moment i am code-blind so pushing this one public for you to debug, check out and provide feedback so we can solve the issue's.

**Comment by Sweepr** [comment]
Currently the installation page will show:

![image](https://user-images.githubusercontent.com/2857865/90643460-ef0c6d80-e233-11ea-8807-25c9e9ef4d66.png)

Where you can fill in the desired `schema`, which i'm not 100% sure it works.

I've tested the install on PostGreSQL using postgresql-13.0-beta3-windows-x64., install went fine, using the `public `schema

**Comment by rogierlommers** [comment]
Hey @Sweepr , thanks for your effort! I tried to install this branch on a clean system. Indeed, the installation screen gives me the possibility to set the schema. When I change this from public into something else, the tables are still being created in the public schema. I also don't see the (custom) schema back in the dbsettings.inc.php file.

**Comment by Sweepr** [comment]
@rogierlommers Thanks for the feedback, i will have a look and figure out how i can improve that.

**Comment by rogierlommers** [comment]
Additional remark: in the database settings screen (during install), I think the phrase `Spotweb database (schema)` should be renamed to `Postgres database`. The other config, `PostGreSQL schema name (Default: public)` should be renamed to `Postgres schema (default: public)`.

**Comment by Sweepr** [comment]
@rogierlommers , true, i'll change that right away.

**Comment by Sweepr** [comment]
@rogierlommers  I have updated the https://github.com/spotweb/spotweb/tree/postgres branch again,

Can you check it out and provide feedback?

**Comment by rogierlommers** [comment]
Started "from scratch" again.
Tried to install with schema; now I get this:

`https://xxxx/install.php?page=99`

<img width="894" alt="Screenshot 2020-08-20 at 15 04 26" src="https://user-images.githubusercontent.com/3907131/90773697-97830600-e2f6-11ea-9d8a-f7d99c3593a9.png">

**Comment by Sweepr** [comment]
Did you create a spotweb database and created the `non-public` schema it should work.

**Comment by rogierlommers** [comment]
You're right! I assumed you would create the schema during install.
But I'm very happy with this change! Perfect! Are you willing to merge into master/release?

**Comment by Sweepr** [comment]
> 
> 
> You're right! I assumed you would create the schema during install.
> But I'm very happy with this change! Perfect! Are you willing to merge into master/release?

Nice!, sure i could have a look and see if i can create the schema during install, i am quite new with PostgreSQL so i was surprised i got this far ;-)  still have to learn how it works, since i am a all-time MySQL user.

I think too soon to merge into Master, first i (and maybe @rogierlommers ) want to use it for a short period of time see if any unexpected issue's pop-up and if we can improve things, then i would first merge it with Develop before i would finally appear in Master, anyway i also need @mesa57 to review and approve this change.

@mesa57 Can you give your green light for this when you have time?

**Comment by rogierlommers** [comment]
Cool, I'll keep on testing your branch. For me there is no real need for the app to create the schema. But then I would like the DB-settings screen to mention something like: `note: the schema should already be present`.

**Comment by Sweepr** [comment]
> 
> 
> Cool, I'll keep on testing your branch. For me there is no real need for the app to create the schema. But then I would like the DB-settings screen to mention something like: `note: the schema should already be present`.

I can do that ;-)

**Comment by mesa57** [comment]
As in mysql database (and even user) is created during install (this was added recently) in PostGreSQL the schema should also be created (if not present already).

**Comment by rogierlommers** [comment]
Here some docs regarding schema creation @ postgres: https://www.postgresql.org/docs/9.3/sql-createschema.html

**Comment by mesa57** [comment]
The 
`CREATE SCHEMA IF NOT EXISTS schema_name [ AUTHORIZATION user_name ]`
seems to to the job :)

**Comment by Sweepr** [comment]
I now made this,

![image](https://user-images.githubusercontent.com/2857865/90788014-9e654500-e305-11ea-902c-4b17934aa278.png)

I let spotweb create the database and schema during install.
This results in 2 schema's:

![image](https://user-images.githubusercontent.com/2857865/90788057-ab823400-e305-11ea-9b9e-6608a7376aae.png)

Is that normal behaviour?

**Comment by Sweepr** [comment]
@rogierlommers 

Added some more changes to https://github.com/spotweb/spotweb/tree/postgres

Include `CREATE SCHEMA IF NOT EXISTS` during install, we weren't sure if 

 `CREATE SCHEMA IF NOT EXISTS schema_name **AUTHORIZATION user_name`** is needed, for now we only use                 `$this->_conn->exec('CREATE SCHEMA IF NOT EXISTS '.$schema.'');` https://github.com/spotweb/spotweb/blob/75677052ead34a2e163107fac6bf012ebf3235db/lib/dbeng/dbeng_pdo_pgsql.php#L30

**Comment by rogierlommers** [comment]
Yup, hereby my confirmation that the latest branch (18:30) works fine. In creates the schema perfect during installation. One thing though: I'm not sure how it works in PHP, but when I look at this:

```
            try {
                $this->_conn = new PDO('pgsql:'.$db_conn.';dbname='.$db, $usr, $pass);
                //$this->_conn->exec('CREATE SCHEMA IF NOT EXISTS '.$schema.' AUTHORIZATION '.$usr.'');
                $this->_conn->exec('CREATE SCHEMA IF NOT EXISTS '.$schema.'');
                $this->_conn->exec('SET search_path TO '.$schema.'');
            } catch (PDOException $e) {
                throw new DatabaseConnectionException($e->getMessage(), -1);
            } // catch
```
Doesn't this mean that for each page view, the code is trying to create the schema?

**Comment by Sweepr** [comment]
For the `CREATE SCHEMA IF NOT EXISTS`,  it will only create the schema if it does not exists.

Not sure about this part tho: $this->_conn->exec('SET search_path TO '.$schema.'');

Anyway for you to see real-time which queries are being executed during normal use?

**Comment by rogierlommers** [comment]
Hey @Sweepr , all statement indeed are idempotent (which means they can safely be executed multiple times). From a functional perspective everything is okay now. But each schema creation (whether or not it exists) takes some CPU cycles. Since the schema creation should only happen once (during the create table-part), it's maybe a good idea to move it to the installation part?

**Comment by Sweepr** [comment]
@rogierlommers

I'll have a look if it can be placed elsewhere, not sure when but probably this weekend, very busy schedule tomorrow. :beers:

**Comment by rogierlommers** [comment]
Same here, 🍺.
Thanks for your help, by the way! (and quick responses)

**Comment by rogierlommers** [comment]
(and let me know if I can help you)

**Comment by mesa57** [comment]
Of course the create schema should only be done on install.
@sweepr : look at the code for mysql. Create database and user are seperate (and I believe I also did that for postgresql). Maybe I have time to look into that tomorrow.

**Comment by Sweepr** [comment]
@mesa57 I tried that but i couldn't get it working quickly, i'll have another look later or if you have time to look at it?
